### ISSUE TITLE
Deprecate "Test" suffix for abstract test case classes

### ISSUE THREAD

**Issue by sebastianbergmann** [sebastianbergmann/phpunit#5132]

**Comment by blackdrago** [comment]
@sebastianbergmann I've been reading through various PRs on Github, the change log, and other PHPUnit documentation, looking for any background or details as to why this change was made and/or what naming convention (if any) is suggested for these abstract test classes, but I haven't found anything about it.

In our testing suite, we have a number of abstract "BaseTest" classes. Some of these establish shared testing setup elements (e.g., loading a specific data set that multiple tests use) while others contain core testing code that's used to test multiple classes (basic code reduction).

Was this change done simply to enforce a requirement (e.g., any class that uses the 'Test' suffix must be a class that PHPUnit can execute directly, and abstract test classes cannot be run directly, so they can't use the suffix), or is there more to it?

I know I'll be asked for 'the why' as soon as I create the ticket for updating the class names, so I figured I'd ask rather than assume.

**Comment by speedbomb** [comment]
@sebastianbergmann I have a question about this deprecation: As far as I can see, this should affect only abstract classes whose class name ends with "Test". There is no such class in our test suite, but we get this error:
```
PHPUnit 9.6.6 by Sebastian Bergmann and contributors.
Warning:       Abstract test case classes with "Test" suffix are deprecated (Tests\Unit\TestCases\BeModuleTestCase)
```
Is this behaviour intended?

**Comment by sebastianbergmann** [comment]
@speedbomb You may have run into an edge case that I will gladly investigate if you can provide a **minimal**, **self-contained**, reproducing test case that shows the problem. Please do so in a new issue. Thanks!

**Comment by speedbomb** [comment]
> @speedbomb You may have run into an edge case [...]

It's conceivable that this edge case is the usage of the IDE _PhpStorm_: The abstract test case `BeModuleTestCase` has 3 inheritors. I can savely run the test method `testCanCallView` for each of them separately, but if i choose to run all 3 of them, the warning occurs.

![Screenshot 2023-05-26 082234](https://github.com/sebastianbergmann/phpunit/assets/5576286/2e932e74-acb8-43a9-8334-2711a42e2b11)

**Comment by Jehong-Ahn** [comment]
`Class Foo` in *Foo_Test.php* causes this problem in my case. 
I made the change to `class Foo_Test`.

**Comment by Mistralys** [comment]
I am currently puzzling over this. On PHPunit 9.6.10 , I get the deprecation message even on classes that do not use any "Text" suffix at all, whether in the class name or the file name e.g. 

`BaseMailcodeChecks.php`

```
Abstract test case classes with "Test" suffix are deprecated (BaseMailcodeChecks)
```

So I had a look at the PHPUnit code that triggers the message, and if I am not mistaken, it means that using `abstract` classes extending `TestCase` are deprecated in general?

`TestSuite.php:459`

```
if ($class->isAbstract() && $class->isSubclassOf(TestCase::class)) {
    $this->addWarning(
        sprintf(
            'Abstract test case classes with "Test" suffix are deprecated (%s)',
            $class->getName(),
        ),
    );
}
```

If that's the case, how should one handle methods common to test classes? With traits?

**Comment by borgogelli** [comment]
> I am currently puzzling over this. On PHPunit 9.6.10 , I get the deprecation message even on classes that do not use any "Test" suffix at all, whether in the class name or the file name e.g.
> 
> `BaseMailcodeChecks.php`
> 
> ```
> Abstract test case classes with "Test" suffix are deprecated (BaseMailcodeChecks)
> ```
> 
> So I had a look at the PHPUnit code that triggers the message, and if I am not mistaken, it means that using `abstract` classes extending `TestCase` are deprecated in general?
> 
> `TestSuite.php:459`
> 
> ```
> if ($class->isAbstract() && $class->isSubclassOf(TestCase::class)) {
>     $this->addWarning(
>         sprintf(
>             'Abstract test case classes with "Test" suffix are deprecated (%s)',
>             $class->getName(),
>         ),
>     );
> }
> ```
> 
> If that's the case, how should one handle methods common to test classes? With traits?

@Mistralys did you find any solutions ? I have the same problem with PHPUnit 10.4.2

**Comment by Mistralys** [comment]
@borgogelli I found out that the error message is only shown when I run a specific test. I am using PHPStorm, and when I choose a single test class or test method to run, I get the message. If I run all tests through `phpunit.xml`, PHPUnit does not complain. Maybe an issue with the scope of the test?

I have come to simply ignore the message.

**Comment by clvarley** [comment]
Sorry to bump an old thread; just posting this for anyone who might be stumbling on this in 2026.

I kept seeing this issue after updating an old project from PHPUnit 9 -> 10, even though my abstract classes followed the correct naming conventions (i.e: `Abstract*TestCase`).

After digging around the PHPUnit source I realised the problem was actually in my `phpunit.xml` config file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<phpunit
    bootstrap="vendor/autoload.php"
    colors="true"
    cacheResult="false"
    testdox="true"
    noInteraction="true"
>
    <testsuites>
        <testsuite name="unit">
            <directory suffix=".php">tests/Unit</directory>
        </testsuite>
    </testsuites>
</phpunit>

```

The [suffix attribute](https://docs.phpunit.de/en/10.5/configuration.html#the-suffix-attribute) defined on `<directory>` was too broad, making PHPUnit consider **every** PHP file in `tests/Unit` to be a test. Simply removing the suffix attribute was enough to resolve the issue for me.

```diff
<testsuite name="unit">
-    <directory suffix=".php">tests/Unit</directory>
+   <directory>tests/Unit</directory>
</testsuite>
```

Leaving it undefined causes the expected behaviour of filtering for `*Test.php` and `*.phpt` files.

I don't remember ever adding the `suffix`, (although this project is 5 years old so I may have done 🤷‍♂️) my current theory is that this was spat out by an older PHPUnit version during `phpunit --generate-configuration`.

**TLDR:** Check both your class names AND config file otherwise PHPUnit might be looking at files you don't want it to.
